File: Binary search tree (BST)/Convert Sorted Array to Binary Search Tree/Convert Sorted Array to Binary Search Tree.cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if(nums.size() == 0) return NULL;
        return BuildBinaryTree(nums, 0 , nums.size() - 1 );
       
    }
    TreeNode* BuildBinaryTree (vector<int>& nums, int left, int right ) {
        if(left > right) return NULL;
        int mid = left + ( right - left ) / 2;
        TreeNode* node = new TreeNode(nums[mid]);
        node->left = BuildBinaryTree(nums, left, mid - 1);
        node->right = BuildBinaryTree(nums, mid + 1, right);
        return node;
    }
};

File: Binary search tree (BST)/Convert Sorted Array to Binary Search Tree/Convert Sorted Array to Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
         if(nums.length == 0) return null;
        return BuildBinaryTree(nums, 0 , nums.length - 1 );
    }
    public TreeNode BuildBinaryTree (int[] nums, int left, int right ) {
        if(left > right) return null;
        int mid = left + ( right - left ) / 2;
        TreeNode node = new TreeNode(nums[mid]);
        node.left = BuildBinaryTree(nums, left, mid - 1);
        node.right = BuildBinaryTree(nums, mid + 1, right);
        return node;
    }
}